CUDA & CPU: support F32 kernel type for `CONV_TRANSPOSE_2D`

AgainstEntropy opened the pull request:
also updated test case in `test-backend-ops`. 
But since F32 kernel type is not supported on CPU, only `GGML_TYPE_F16` is kept and `GGML_TYPE_F32` can be uncommented back in the future.

AgainstEntropy committed:
Refactor CUDA 2D transpose implementation to support multiple kernel types and improve parameter handling

- Introduced a `conv2d_transpose_params` struct for better parameter management.
- Updated `conv2d_transpose_kernel` to be templated for different kernel types (float and half).
- Modified `ggml_cuda_conv_2d_transpose_p0` to handle both F16 and F32 kernel types.
- Enhanced test cases to validate functionality for both kernel types.

AgainstEntropy committed:
Refactor test cases for 2D convolution transpose to support dynamic kernel types

- Updated `test_conv_transpose_2d` structure to improve parameter handling by reordering constructor arguments.
- Enhanced test case generation to iterate over kernel types, allowing for flexible testing of different configurations.
- Removed hardcoded kernel type instances in favor of a loop for better maintainability and scalability.

am17an reviewed (commented):
Does this PR make a difference to something? From what I understand, the kernel value is upcast into float before doing any accumulation (and accumulation is anyway in f32). So unless there are kernels around which don't fit into f16 I don't see a benefit to supporting this, especially when we don't support the f16 inputs yet (which incidentally might be more relevant than kernels being f32 as we could potentially do `half2` multiplications)

AgainstEntropy commented on ggml/src/ggml-cuda/conv2d-transpose.cu:
```diff
@@ -83,9 +143,12 @@ void ggml_cuda_conv_2d_transpose_p0(ggml_backend_cuda_context & ctx, ggml_tensor
-        input_data, kernel_data, output_data, input_w, input_h, output_w, output_h, kernel_w, kernel_h, stride,
-        channels_in, channels_out, batches);
+    if (kernel->type == GGML_TYPE_F16) {
+        conv2d_transpose_cuda_f16(input_data, (const half *) kernel_data, output_data, params, st);
```
I referred to `conv2d.cu` for the current dispatching manner, and I thought there is some convention in llama.cpp 😂

https://github.com/ggml-org/llama.cpp/blob/8e878f0cb4c893de23455dd0a6bfbbb21bcaad89/ggml/src/ggml-cuda/conv2d.cu#L108-L120

https://github.com/ggml-org/llama.cpp/blob/8e878f0cb4c893de23455dd0a6bfbbb21bcaad89/ggml/src/ggml-cuda/conv2d.cu#L161-L165

am17an replied:
It's not a convention and I would prefer to avoid 1 line functions

AgainstEntropy committed:
Refactor ggml_compute_forward_conv_transpose_2d to support both F16 and F32 tensor types.

AgainstEntropy committed:
Refactor conv2d transpose kernel to use a template for kernel type, enhancing flexibility for different data types.
Update test cases to include both F16 and F32 tensor types for comprehensive coverage.

am17an commented on tests/test-backend-ops.cpp:
```diff
@@ -4487,28 +4487,33 @@ struct test_conv_transpose_1d : public test_case {
     const std::array<int64_t, 4> ne_kernel;
     const int stride;
+    // Types
+    const ggml_type type_kernel;
```
kernel_type is a more conventional name

AgainstEntropy committed:
Update ggml/src/ggml-cuda/conv2d-transpose.cu

Co-authored-by: Aman Gupta <amangupta052@gmail.com>

AgainstEntropy committed:
Update ggml/src/ggml-cpu/ggml-cpu.c

Co-authored-by: Aman Gupta <amangupta052@gmail.com>

AgainstEntropy committed:
Refactor conv2d transpose implementation by removing the conv2d_transpose_params struct and dispatching with direct kernel launch.

AgainstEntropy committed:
Enhance cpu conv2d transpose implementation by introducing a templated kernel type for improved flexibility with F16 and F32 data types.

am17an reviewed (approved):
I'm okay with the CUDA changes. Either @ggerganov or @slaren has to approve the ggml-cpu changes.

am17an reviewed (approved):
Yes still looks good. Need to rebase
